Add Multi Datacenter ScyllaDB cluster deployment documentation (#1466)
scylladb/scylla-operator: docs/source/multidc/multidc.md

                operator: In
                values:
                - scylla
        tolerations:
        - key: role
          operator: Equal
          value: scylla-clusters
          effect: NoSchedule
    - name: c
      members: 1
      storage:
        storageClassName: scylladb-local-xfs
        capacity: 1800G
      resources:
        limits:
          cpu: 7
          memory: 56G
      placement:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: topology.kubernetes.io/zone
                operator: In
                values:
                - us-east-1c
              - key: scylla.scylladb.com/node-type
                operator: In
                values:
                - scylla
        tolerations:
        - key: role
          operator: Equal
          value: scylla-clusters
          effect: NoSchedule
```

Apply the manifest:
```console
kubectl --context "${CONTEXT_DC1}" apply --server-side -f dc1.yaml
```

Wait for to the cluster to be fully rolled out and all of its nodes to be in state UN. You can verify that by running: 
```console

> kubectl --context "${CONTEXT_DC1}" -n scylla exec -it pod/scylla-cluster-us-east-1-a-0 -c scylla -- nodetool status

Reviewer: Lets wait for condition, and then check nodetool status

```

The expected output should look similar to the below:
```console
Datacenter: us-east-1
=====================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address      Load       Tokens       Owns    Host ID                               Rack
UN  10.0.70.195  290 KB     256          ?       494277b9-121c-4af9-bd63-3d0a7b9305f7  c
UN  10.0.59.24   559 KB     256          ?       a3a98e08-0dfd-4a25-a96a-c5ab2f47eb37  b
UN  10.0.19.237  107 KB     256          ?       64b6292a-327f-4128-852a-6004039f402e  a
```

Use the below commands and their expected outputs as a reference for retrieving the PodIPs used by the cluster for inter-node communication.
```console
kubectl --context "${CONTEXT_DC1}" -n scylla get pod/scylla-cluster-us-east-1-a-0 -ogo-template --template '{{ .status.podIP }}'
10.0.19.237

kubectl --context "${CONTEXT_DC1}" -n scylla get pod/scylla-cluster-us-east-1-b-0 -ogo-template --template '{{ .status.podIP }}'
10.0.59.24

kubectl --context "${CONTEXT_DC1}" -n scylla get pod/scylla-cluster-us-east-1-c-0 -ogo-template --template '{{ .status.podIP }}'
10.0.70.195
```

> We are going to utilize the retrieved addresses as seeds for the other datacenter.

Reviewer: We should make a note here that PodIPs are ephemeral, so users should consider using DNS that would make sure they are up to date. You can mention you skipped it for the simplicity of this document.

#### Deploy the second datacenter

To deploy the second datacenter, you will follow similar steps.

First, create a dedicated 'scylla' namespace:
```console
kubectl --context "${CONTEXT_DC2}" create ns scylla
```

Replace the values in `.spec.externalSeeds` of the below manifest with the Pod IP addresses that you have retrieved earlier.
The provided values are going to serve as initial contact points for the joining nodes of the second datacenter.

For the purpose of this guide, we are assuming that the second cluster is running in `us-east-2` region and the nodes dedicated for running ScyllaDB nodes are running in zones `us-east-2a`, `us-east-2b` and `us-east-2c` correspondingly. If that is not the case, adjust the manifest accordingly.
Having configured it, save the manifest as `dc2.yaml`:

> ```yaml

Reviewer: You can mention that name of the ScyllaCluster must be in sync for all datacenters.

                values:
                - us-east-2b
              - key: scylla.scylladb.com/node-type
                operator: In
                values:
                - scylla
        tolerations:
        - key: role
          operator: Equal
          value: scylla-clusters
          effect: NoSchedule
    - name: c
      members: 1
      storage:
        storageClassName: scylladb-local-xfs
        capacity: 1800G
      resources:
        limits:
          cpu: 7
          memory: 56G
      placement:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: topology.kubernetes.io/zone
                operator: In
                values:
                - us-east-2c
              - key: scylla.scylladb.com/node-type
                operator: In
                values:
                - scylla
        tolerations:
        - key: role
          operator: Equal
          value: scylla-clusters
          effect: NoSchedule
```

To apply the manifest, run:
```console
kubectl --context "${CONTEXT_DC2}" -n scylla apply --server-side -f dc2.yaml
```

> Wait for the second datacenter to roll out. You can verify that the nodes have joined the existing cluster and that you are now running a multi datacenter ScyllaDB cluster by running `nodetool status` with the below command:

Reviewer: Add a command user should use to wait for rollout.

            nodeSelectorTerms:
            - matchExpressions:
              - key: topology.kubernetes.io/zone
                operator: In
                values:
                - us-east-1b
              - key: scylla.scylladb.com/node-type
                operator: In
                values:
                - scylla
        tolerations:
        - key: role
          operator: Equal
          value: scylla-clusters
          effect: NoSchedule
    - name: c
      members: 1
      storage:
        storageClassName: scylladb-local-xfs
        capacity: 1800G
      resources:
        limits:
          cpu: 7
          memory: 56G
      placement:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: topology.kubernetes.io/zone
                operator: In
                values:
                - us-east-1c
              - key: scylla.scylladb.com/node-type
                operator: In
                values:
                - scylla
        tolerations:
        - key: role
          operator: Equal
          value: scylla-clusters
          effect: NoSchedule
```

> Note, that

Reviewer: Looks like something is missing

Author: It got there by mistake, thanks!

```
```console
scyllacluster.scylla.scylladb.com/scylla-cluster condition met
```

You can now verify that all the nodes of your cluster are in UN state:
```shell
kubectl --context="${CONTEXT_DC1}" -n=scylla exec -it pod/scylla-cluster-us-east-1-a-0 -c=scylla -- nodetool status
```

The expected output should look similar to the below:
```console
Datacenter: us-east-1
=====================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address      Load       Tokens       Owns    Host ID                               Rack
UN  10.0.70.195  290 KB     256          ?       494277b9-121c-4af9-bd63-3d0a7b9305f7  c
UN  10.0.59.24   559 KB     256          ?       a3a98e08-0dfd-4a25-a96a-c5ab2f47eb37  b
UN  10.0.19.237  107 KB     256          ?       64b6292a-327f-4128-852a-6004039f402e  a
```

##### Retrieve PodIPs of ScyllaDB nodes for use as external seeds

``` warning::
    Due to the ephemeral nature of PodIPs, it is ill-advised to use them as seeds in production environments.
    PodIPs are being used in this example for the sheer simplicity of this setup.

>     In production environments, it is recommended that you use domain names or non-ephemeral IP addresses.

Reviewer: ```suggestion
    In production environments, it is recommended that you use domain names.
```

Sounds like we discourage from using PodIPs, but we don't.

Author: we do for external seeds though, don't we?

Reviewer: right it's mentioned in seed context. 
Anyway, this warning could use describing what's the problem of using ephemeral ones, which is that they in future may no longer point to nodes from this ScyllaCluster. To solve it, we recommend setting up DNS which would resolve to ephemeral IPs, or somehow provide ScyllaCluster Pods with static IP address and use them as seeds.

Author: done

```

Use the below commands and their expected outputs as a reference for retrieving the PodIPs used by the cluster for inter-node communication.
```shell

> kubectl --context="${CONTEXT_DC1}" -n=scylla get pod/scylla-cluster-us-east-1-a-0 -o=go-template --template='{{ .status.podIP }}'

Reviewer 2: ```suggestion
kubectl --context="${CONTEXT_DC1}" -n=scylla get pod/scylla-cluster-us-east-1-a-0 --template='{{ .status.podIP }}'
```
(globally)

To deploy the second datacenter, you will follow similar steps.

First, create a dedicated 'scylla' namespace:
```shell
kubectl --context="${CONTEXT_DC2}" create ns scylla
```

Replace the values in `.spec.externalSeeds` of the below manifest with the Pod IP addresses that you retrieved earlier.
The provided values are going to serve as initial contact points for the joining nodes of the second datacenter.

For this guide, let's assume that the second cluster is running in `us-east-2` region and the nodes dedicated for running ScyllaDB nodes are running in zones `us-east-2a`, `us-east-2b` and `us-east-2c` correspondingly. If that is not the case, adjust the manifest accordingly.
Having configured it, save the manifest as `dc2.yaml`:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
metadata:
  name: scylla-cluster
  namespace: scylla
spec:
  version: 5.2.7
  agentVersion: 3.1.2
  cpuset: true
  exposeOptions:
    broadcastOptions:
      clients:
        type: PodIP
      nodes:
        type: PodIP
    nodeService:
      type: Headless
  externalSeeds:
  - 10.0.19.237
  - 10.0.59.24
  - 10.0.70.195
  datacenter:
    name: us-east-2
    racks:

>     - name: a

Reviewer: specify `agentResources` to ensure Guaranteed QoS (in all examples)

Author: done

- running ScyllaDB Operator and its prerequisites
- running a storage provisioner capable of provisioning XFS volumes of StorageClass `scylladb-local-xfs` in each of the nodes dedicated to ScyllaDB instances

You can refer to one of our guides describing the process of preparing such infrastructure:
- [Build multiple Amazon EKS clusters with Inter-Kubernetes networking](eks.md)
- [Build multiple GKE clusters with Inter-Kubernetes networking](gke.md)

Additionally, to follow the below guide, you need to install and configure the following tools that you will need to manage Kubernetes resources:
- kubectl – A command line tool for working with Kubernetes clusters.

See [Install Tools](https://kubernetes.io/docs/tasks/tools/) in Kubernetes documentation for reference.

## Multi Datacenter ScyllaDB Cluster

In v1.11, ScyllaDB Operator introduced support for manual multi-datacenter ScyllaDB cluster deployments.

``` warning::
    ScyllaDB Operator only supports *manual configuration* of multi-datacenter ScyllaDB clusters.

>     In other words, although ScyllaCluster API exposes the machinery necessary for setting up multi-datacenter ScylaDB clusters, the ScyllaDB Operator only automates operations related to singular datacenters.

Reviewer 2: ```suggestion
    In other words, although ScyllaCluster API exposes the machinery necessary for setting up multi-datacenter ScylaDB clusters, the ScyllaDB Operator only automates operations for a single datacenter.
```

Author: done

    Operations related to multiple datacenters may require manual intervention of a human operator.
    Most notably, destroying one of the Kubernetes clusters or ScyllaDB datacenters is going to leave DN nodes behind in other datacenters, and their removal has to be carried out manually.
```

The main mechanism used to set up a manual multi-datacenter ScyllaDB cluster is a field in ScyllaCluster's specification - `externalSeeds`.

### External seeds

> The `externalSeeds` field in ScyllaCluster's specification enables control over external seeds propagated to ScyllaDB binary as `seeds` parameter of seed-provider.

Reviewer 2: ```suggestion
The `externalSeeds` field in ScyllaCluster's specification enables control over external seeds that are propagated to ScyllaDB binary using additional `seeds` arguments.
```

Author: seeds-provider is a config options in ScyllaDB config, so this suggestion removes a bit of accuracy - was that intended?

Reviewer 2: i thought we set the `seed` flag, not mess with the config

Author: To the docker entrypoint, yes, but that ends up being provided as `--seed-provider-parameters seeds=<seeds>`

Author: done

> External in this context should be understood as external to the datacenter being specified by the API.

Reviewer 2: ```suggestion
External, in this context, should be understood as external to the datacenter being specified by the API.
```

Author: changed a bit but done

      clients:
        type: PodIP
      nodes:
        type: PodIP
```

However, other configuration options allow for the manual deployment of multi-datacenter ScyllaDB clusters in different network setups. For details, refer to [Exposing ScyllaClusters](../exposing.md) in ScyllaDB Operator documentation.

#### Deploy a multi-datacenter ScyllaDB Cluster

#### Using context

We are going to specify contexts for `kubectl` commands used throughout the guide.
To retrieve the context of your current cluster, run:
```shell
kubectl config current-context
```

Save the contexts of the two clusters, which we are going to deploy the datacenters in, as `CONTEXT_DC1` and `CONTEXT_DC2` environment variables correspondingly.

#### Deploy the first datacenter

First, run the below command to create a dedicated 'scylla' namespace:
```shell
kubectl --context "${CONTEXT_DC1}" create ns scylla
```

For this guide, we are assuming that your cluster is running in `us-east-1` region and the nodes dedicated to running ScyllaDB nodes are running in zones `us-east-1a`, `us-east-1b` and `us-east-1c` correspondingly. If that is not the case, adjust the manifest accordingly.

``` caution::

>     The ``.spec.name`` fields of the ScyllaCluster objects have to be consistent across all datacenters being deployed.

Reviewer 2: ```suggestion
    The ``.spec.name`` fields of the ScyllaCluster objects represent the ScyllaDB cluster name and has to match between all datacenters of this ScyllaDB cluster.
```

Author: done

    The names of the datacenters, specified in ``.spec.datacenter.name``, have to be unique across the entire multi-datacenter cluster. \
    
    
    For more information see `Create a ScyllaDB Cluster - Multi Data Centers (DC) <https://opensource.docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/create-cluster-multidc.html>`_ in ScyllaDB documentation.
```

Save the ScyllaCluster manifest in `dc1.yaml`:
```yaml
apiVersion: scylla.scylladb.com/v1
kind: ScyllaCluster
metadata:
  name: scylla-cluster
  namespace: scylla
spec:
  version: 5.2.7
  agentVersion: 3.1.2
  cpuset: true
  exposeOptions:
    broadcastOptions:
      clients:
        type: PodIP
      nodes:
        type: PodIP
    nodeService:
      type: Headless
  datacenter:
    name: us-east-1
    racks:
    - name: a
      members: 1
      storage:
        storageClassName: scylladb-local-xfs
        capacity: 1800G
      resources:
        limits:
          cpu: 7
          memory: 56G
      placement:

>         nodeAffinity:

Reviewer 2: for prod, we also need anti-affinity so they don't land on the same node with 2+ members

Reviewer 2: I guess it's not fatal with racks but still better to spread them, at least as `preferredDuringSchedulingIgnoredDuringExecution` if not required

Reviewer 2: (using `topologyKey: kubernetes.io/hostname`)

Author: should this be covered in this doc (this manifest specifically) or in the general guide?

Reviewer 2: in every manifest and you have one here. I suppose it's more a question whether we keep piling up the deficiency or start fixing them at least in new manifests.

Author: Yup, but that's a more extensive topic, which I suppose is out of scope for this doc. I'll fix it in the new manifests

Author: done

              - key: topology.kubernetes.io/zone
                operator: In
                values:
                - us-east-1b
              - key: scylla.scylladb.com/node-type
                operator: In
                values:
                - scylla
        tolerations:
        - key: role
          operator: Equal
          value: scylla-clusters
          effect: NoSchedule
    - name: c
      members: 1
      storage:
        storageClassName: scylladb-local-xfs
        capacity: 1800G
      resources:
        limits:
          cpu: 7
          memory: 56G
      placement:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: topology.kubernetes.io/zone
                operator: In
                values:
                - us-east-1c
              - key: scylla.scylladb.com/node-type
                operator: In
                values:
                - scylla
        tolerations:
        - key: role
          operator: Equal
          value: scylla-clusters
          effect: NoSchedule
```

Note, that

Apply the manifest:
```shell

> kubectl --context "${CONTEXT_DC1}" apply --server-side -f dc1.yaml

Reviewer 2: --flag=value (applies globally)

Author: done

```

The expected output should look similar to the below:
```console
Datacenter: us-east-1
=====================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address      Load       Tokens       Owns    Host ID                               Rack
UN  10.0.70.195  290 KB     256          ?       494277b9-121c-4af9-bd63-3d0a7b9305f7  c
UN  10.0.59.24   559 KB     256          ?       a3a98e08-0dfd-4a25-a96a-c5ab2f47eb37  b
UN  10.0.19.237  107 KB     256          ?       64b6292a-327f-4128-852a-6004039f402e  a
```

##### Retrieve PodIPs of ScyllaDB nodes for use as external seeds

``` warning::
    Due to the ephemeral nature of PodIPs, it is ill-advised to use them as seeds in production environments. 
    This is because there is a high likelihood that the Pods of your ScyllaDB clusters will change their IPs during the cluster's lifecycle, and so the provided seeds will no longer point to the ScyllaDB nodes.
    It is undesired, as the seeds provided on node's startup may serve as fallback contact points when all of the node's peers are unreachable.
    PodIPs are being used in this example for the sheer simplicity of this setup.

>     In production environments, it is recommended that you use domain names or non-ephemeral IP addresses as external seeds.

Reviewer: I would swap these two sententces. First recommend, then explain why we are not following the recommendation.

Author: done
